File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int maxFruit = 0;
        int start = 0;
        unordered_map<int, int> basket;
    
        for (int end = 0; end < fruits.size(); end++) {
            basket[fruits[end]]++;
    
            while (basket.size() > 2) {
                basket[fruits[start]]--;
                if (basket[fruits[start]] == 0) {
                    basket.erase(fruits[start]);
                }
                start++;
            }
    
            maxFruit = max(maxFruit, end - start + 1);
        }
    
        return maxFruit;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cs
public class Solution {
    public int TotalFruit(int[] fruits) {
        int maxFruit = 0;
        int start = 0;
        Dictionary<int, int> basket = new Dictionary<int, int>();

        for (int end = 0; end < fruits.Length; end++) {
            if (basket.ContainsKey(fruits[end])) {
                basket[fruits[end]]++;
            } else {
                basket[fruits[end]] = 1;
            }

            while (basket.Count > 2) {
                basket[fruits[start]]--;
                if (basket[fruits[start]] == 0) {
                    basket.Remove(fruits[start]);
                }
                start++;
            }

            maxFruit = Math.Max(maxFruit, end - start + 1);
        }

        return maxFruit;
    }
}